nutils/elementseq.py

```diff
@@ -229,159 +321,108 @@ class UniformReferences(References):
     self._length = length
-    super().__init__(reference.ndims)
+    super().__init__(item.ndims)
 
   def __len__(self):
     return self._length
 
-  def __getitem__(self, index):
-    if numeric.isint(index):
-      numeric.normdim(len(self), index)
-      return self._reference
-    elif isinstance(index, slice):
-      return asreferences([self._reference], self.ndims) * len(range(len(self))[index])
-    elif numeric.isintarray(index) and index.ndim == 1:
-      if numpy.any(numpy.less(index, 0)) or numpy.any(numpy.greater_equal(index, len(self))):
-        raise IndexError('index out of range')
-      return asreferences([self._reference], self.ndims) * len(index)
-    elif numeric.isboolarray(index) and index.shape == (len(self),):
-      return asreferences([self._reference], self.ndims) * numpy.sum(index)
+  def __iter__(self):
+    return itertools.repeat(self._item, len(self))
+
+  def get(self, index):
+    numeric.normdim(len(self), index)
+    return self._item
+
+  def take(self, indices):
+    _checktake(len(self), indices)
+    return Uniform(self._item, 1).repeat(len(indices))
+
+  def compress(self, mask):
+    _checkcompress(len(self), mask)
+    return Uniform(self._item, 1).repeat(mask.sum())
+
+  def repeat(self, count):
+    if count == 0:
+      return Empty(self.ndims)
     else:
-      return super().__getitem__(index)
+      return Uniform(self._item, len(self) * count)
+
+  def product(self, other):
+    if other.isuniform:
+      return Uniform(self._item * other.get(0), len(self) * len(other))
+    else:
+      return super().product(other)
 
   @property
   def children(self):
-    return asreferences(self._reference.child_refs, self.ndims) * len(self)
+    return asreferences(self._item.child_refs, self.ndims).repeat(len(self))
 
   @property
   def edges(self):
-    return asreferences(self._reference.edge_refs, self.ndims-1) * len(self)
-
-  def getpoints(self, ischeme, degree):
-    return (self._reference.getpoints(ischeme, degree),)*len(self)
-
-  def __mul__(self, other):
-    if numeric.isint(other):
-      if other == 0:
-        return EmptyReferences(self.ndims)
-      else:
-        return UniformReferences(self._reference, len(self)*other)
-    else:
-      return super().__mul__(other)
+    return asreferences(self._item.edge_refs, self.ndims-1).repeat(len(self))
 
   @property
   def isuniform(self):
-    return True
+    return bool(self)
+
+  def getpoints(self, ischeme, degree):
+    return pointsseq.Uniform(self._item.getpoints(ischeme, degree), len(self))
 
-class SelectedReferences(References):
+class Take(References):
   '''A selection of references.  Duplication and reordering is allowed.
 
   Parameters
   ----------
   parent : :class:`References`
-      The transforms to subset.
+      The sequence to select from.
   indices : one-dimensional array of :class:`int`\\s
-      Indices of ``parent`` that form this selection.
+      The indices of ``parent`` references to select.
   '''
 
```

Comment:
is it correct that Take does not define getpoints?

Reply:
This class is probably only used with `_Plain` as parent, in which case `parent.getpoints(...).take(...)` is less efficient than the base implementation.